test/mscclpp-test/allgather_test.cu

```diff
@@ -288,6 +288,49 @@ __global__ void allgather4(int rank, int worldSize, int nRanksPerNode, size_t ne
                    nBlocksForLocalAllGather);
 }
 
+__global__ void allgather5(int rank, int worldSize, int nRanksPerNode, size_t nelemsPerGPU) {
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  const int lid = tid % WARP_SIZE;
+  const int wid = tid / WARP_SIZE;
+  const int nWarp = blockDim.x * gridDim.x / WARP_SIZE;
+  const int nPeer = nRanksPerNode - 1;
+  const int chanOffset = nPeer * blockIdx.x;
+  auto smChans = constSmChans + chanOffset;
+
+  if (wid < nPeer) {
+    smChans[wid].signal();
+    smChans[wid].wait();
+  }
+  __syncthreads();
+  constexpr size_t unitBytesPerThread = 16;
+  constexpr size_t unitBytesPerWarp = unitBytesPerThread * WARP_SIZE;
+  const size_t unitBytes = unitBytesPerWarp * nWarp;
+  const size_t bytesPerGPU = nelemsPerGPU * sizeof(int);
+  const size_t bytes = bytesPerGPU * nPeer;
+  const size_t nLoop = bytes / unitBytes;
+  for (size_t i = 0; i < nLoop; ++i) {
+    const size_t gWid = wid + i * nWarp;
+    const int peerIdx = gWid % nPeer;
+    const int remoteRankLocalIndex = (peerIdx < rank ? peerIdx : peerIdx + 1);
+    const size_t offset = bytesPerGPU * remoteRankLocalIndex + (gWid / nPeer) * unitBytesPerWarp;
+    smChans[peerIdx].get(offset, unitBytesPerWarp, lid, WARP_SIZE);
```

Comment:
I am a bit confused about this algo. For example, a thread block has 16 warps. Then the peerIdx is: "0,1,2,3,4,5,6,0,1,2,3,4,5,6,0,1". As you see they will get 3 pieces of data from peer `0,1`, and two pieces of data from `"0,1,2,3,4,5,6`. Seems not correct.